src/rules.jl

```diff
@@ -96,3 +96,110 @@ true
 See also: [`frule`](@ref), [`@scalar_rule`](@ref)
 """
 rrule(::Any, ::Vararg{Any}; kwargs...) = nothing
+
+
+#######################################################################
+# Infastructure to support generating overloads from rules.
+
+const NEW_RRULE_HOOKS = Function[]
+const NEW_FRULE_HOOKS = Function[]
+_hook_list(::typeof(rrule)) = NEW_RRULE_HOOKS
+_hook_list(::typeof(frule)) = NEW_FRULE_HOOKS
+
+"""
+    on_new_rule(sig->eval...), frule | rrule)
+"""
+function on_new_rule(hook_fun, rule_kind)
+    # get all the existing rules
+    ret = map(_rule_list(rule_kind)) do method
+        sig = _primal_sig(rule_kind, method)
+        _safe_hook_fun(hook_fun, sig)
+    end
+
+    # register hook for new rules
+    push!(_hook_list(rule_kind), hook_fun)
+    return ret
+end
+
+function __init__()
+    push!(Base.package_callbacks, pkgid -> refresh_rules())
```

Comment:
Perhaps add a comment, since these are relatively internal Base things?
 e.g.
```suggestion
    # Need to refresh rules when a module is loaded or a file is `include`d.
    push!(Base.package_callbacks, pkgid -> refresh_rules())
```
I know this kinda says what the code does (not why), but might still be worth it in this case (to save people having to go look in Base to see what these things are).


(out of scope here, but I wonder if a PR to Base to turn those comments into docstrings - so `? Base.package_callbacks` is informative - would be accepted.)